lib/pengine/utils.c

```diff
@@ -1263,9 +1264,28 @@ unpack_operation(pe_action_t * action, xmlNode * xml_obj, pe_resource_t * contai
         }
     }
 
-    value = g_hash_table_lookup(action->meta, XML_ATTR_TIMEOUT);
+    value = NULL;
+    if (pcmk__str_any_of(action->task, RSC_START, RSC_STATUS, NULL)
+        && action->rsc->parameters && action->rsc->xml
+        && safe_str_eq(crm_element_value(action->rsc->xml,
+                                         XML_AGENT_ATTR_CLASS),
+                       PCMK_RESOURCE_CLASS_STONITH)) {
```

Comment:
This is an unusual case since pcmk_monitor_timeout is so closely tied to stonith resources, but looking at it generally, we recently moved to abstracting class-specific comparisons like this, and it's probably a good idea for future-proofing and consistency to adopt that here.

The pcmk_ra_caps enum from include/crm/common/util.h defines all the things that can vary by resource class. The pcmk_get_ra_caps() function from lib/common/agents.c returns a mask of those, given the class name. Calling code can test that mask using the usual bitwise macros to ask "does this class have this capability".

The idea is that, if we one day add a new resource class, we simply add it to pcmk_get_ra_caps() and no other code needs updating.

Reply:
> it's probably a good idea for future-proofing and consistency to adopt that here.

What would that look like? Would we need to define a new pcmk_ra_cap for what we're trying to do with pcmk_monitor_timeout?

I don't see what existing capability we should be checking for here in lieu of checking whether the class is stonith. I'm assuming you're **not** talking about checking pcmk_ra_cap_params instead of action->rsc->parameters, as the cap check wouldn't replace the NULL check here. The other two stonith caps (unique and stdin) seem irrelevant in this context.

Reply:
Correct, this would be a new capability for "support pcmk_monitor_timeout, etc. params". No thoughts on a snappy name for that.

The idea is to untie capability handling from specific classes. Instead of "handle these classes specially", we get "classes have these capabilities, handle each capability like this".

For example if someday a new fence agent standard is created that's just a little different from the stonith class, we can just specify any overlapping capabilities and all existing code will pick it up without changes. Of course if it offers completely new functionality it will need new capabilities and handling, but that's the fun part compared to adding "|| safe_str_eq(class, ...)" in a dozen places.